### DISCUSSION TITLE
Dynamic rules for property binding

### DISCUSSION THREAD

**Discussion by tmrbaset** [livewire/livewire#2569] [1. Help]
I am building a Livewire component that I intend to use dynamically with different models. It will simply display a table with all records of a certain models, with buttons to delete and edit.

My aim is to just pass the model class to the livewire component when it's called in the blade file, and the component will parse its fields, create the table, etc...

Everything worked fine until I tried to add/edit records. the way I handled it is to set a `$selectedItem` when the edit button is clicked, using its ID, I show a dialog with input fields based on the fields of the Model, which are passed by a `const FIELDS` from the class. and bind the properties to the `$selectedItem` in the Livewire component. I declare a `protected $rules` in the component in order to be able to bind the properties. the rules are also defined in the FIELDS constant in the model itself.

here is an example:

```
class Person extends Model
{
    const FIELDS = [
        'name' => 'required|string',
        'email' => 'required|email'
    ] ;
}
```

in the component:

 ```
   public $model;
    public $model_name;
    public $model_fields;
   protected $rules;
   /// other variables

    public function mount($model, $model_name)
    {
        $this->model = $model;
        $this->setRules();  //set the rules from the model to the protected variable,

        $this->model_name = $model_name;
        $this->model_fields = array_keys($model::FIELDS);
    }

    private function setRules(){
        $rules = $this->model::FIELDS;

        foreach($rules as $key=>$value){
            $this->rules['selectedItem.'.$key] = $value;            
        }
    }


```

in the blade file, the fields are defined like this:

```
            @foreach($model_fields as $field)
                        <div>
                            <label class="text-gray-700" for="{{$field}}">{{str_replace("_", " ", $field)}}</label>
                            <input
                                class="someClass" type="text" 
                                 id="{{$field}}" wire:model="selectedItem.{{$field}}" wire:dirty.class="border-red-500">
                        </div>
              @endforeach
```

So esseitially I set the rules dynamically based on the fields passed from the model class. unfortunately this doesn't work. I always get this kind of errors, as if the $rules are not defined:

```
Livewire\Exceptions\CannotBindToModelDataWithoutValidationRuleException
Cannot bind property [selectedItem.name] without a validation rule present in the [$rules] array on Livewire component: [list-editor].
```

I tried to define the rules in the ` protected function rules()` as well. but no difference. 

If, however, I define the $rules array directly in the component in an array, everything works fine. I tried even to change it dynamically, but it still holds up the initial array, it doesn't change when mount() is called.

The question is: Is it possible to define validation rules (mainly for property binding) dynamically in Livewire components? if so, what am I doing wrong?

**Comment by joshhanley** [comment]
@tmrbaset mount function only runs on initial request, so as you are calling `$this->setRules()` in there, that won't be called again on a subsequent request, as such the rules will be empty.

You should create a hydrate method, and also call `$this->setRules()` in there and that should work. https://laravel-livewire.com/docs/2.x/lifecycle-hooks

Hope this helps!

**Reply by tmrbaset** [reply]
Thanks for replying.

Why would  the rules be empty? If `$this->setRules()` is called in mount(), shouldn't this set the rules on mount? I need it to be set only once, no need to reset it again.

**Reply by joshhanley** [reply]
@tmrbaset when a Livewire request happens, it has to re-apply everything including the rules. Normally, if the rules are just a static array on the class, they will be there each time the class in instantiated.

You are trying to do it dynamically though, so as the mount method doesn't run on subsequent requests, the rules will be empty. That is why you are getting that error, as there are no rules applied on subsequent requests. Hence why you need to add it in the hydrate method. 

Hope that makes sense

**Reply by tmrbaset** [reply]
yes, makes sense now, It seems I have to rethink how to do this.

Thanks a lot for your help.
